### ISSUE TITLE
Dataset problem: Can't concatenate CESM2 historical and ssp245 because of different bounds_dtype in latitude and longitude.

### ISSUE THREAD

**Issue by npgillett** [ESMValGroup/ESMValCore#576]
CESM2 monthly mean tas historical has `bounds_dtype float32` and corresponding ssp245 file has `bounds_dtype float64`. I want to concatenate historical and ssp245 and this gives an error when trying to concatenate cubes. I tried to write a fix for this model, but apparently this attribute is read-only, so it's not possible simply to overwrite one of the attributes. How should I proceed?

- Project (CMIP5/CMIP6/OBS/obs4mips/etc): CMIP6
- Full dataset description (fill out as many as you know, please):
  - experiment: historical, ssp245
  - mip: cmip, scenariomip
  - grid: gn
  - version: See below for filepath.
  - variable used: tas
- Problems encountered (please describe what the actual problems are: e.g. wrong standard name, issue with dimensions):
Fails with this message:
```
  File "/home/rng/miniconda3/envs/esmvaltool2/lib/python3.7/site-packages/iris/_concatenate.py", line 480, in match
    raise iris.exceptions.ConcatenateError(msgs)
iris.exceptions.ConcatenateError: failed to concatenate into a single cube.
  Dimension coordinates metadata differ: latitude, longitude != latitude, longitude
```
The `dim_metadata` which are compared and which lead to the above failure are the following:
```
**** 1 [_CoordMetaData(defn=CoordDefn(standard_name='time', long_name=None, var_name='time', units=Unit('days since 1850-1-1 00:00:00', calendar='365_day'), attributes={'title': 'time', 'type': 'double'}, coord_system=None), dims=(0,), points_dtype=dtype('float64'), bounds_dtype=dtype('float64'), kwargs={'scalar': False, 'circular': False, 'order': 1}), _CoordMetaData(defn=CoordDefn(standard_name='latitude', long_name=None, var_name='lat', units=Unit('degrees_north'), attributes={'title': 'Latitude', 'type': 'double', 'valid_max': 90.0, 'valid_min': -90.0}, coord_system=None), dims=(1,), points_dtype=dtype('float64'), bounds_dtype=dtype('float32'), kwargs={'scalar': False, 'circular': False, 'order': 1}), _CoordMetaData(defn=CoordDefn(standard_name='longitude', long_name=None, var_name='lon', units=Unit('degrees_east'), attributes={'title': 'Longitude', 'type': 'double', 'valid_max': 360.0, 'valid_min': 0.0}, coord_system=None), dims=(2,), points_dtype=dtype('float64'), bounds_dtype=dtype('float32'), kwargs={'scalar': False, 'circular': True, 'order': 1})]

**** 2 [_CoordMetaData(defn=CoordDefn(standard_name='time', long_name=None, var_name='time', units=Unit('days since 1850-1-1 00:00:00', calendar='365_day'), attributes={'title': 'time', 'type': 'double'}, coord_system=None), dims=(0,), points_dtype=dtype('float64'), bounds_dtype=dtype('float64'), kwargs={'scalar': False, 'circular': False, 'order': 1}), _CoordMetaData(defn=CoordDefn(standard_name='latitude', long_name=None, var_name='lat', units=Unit('degrees_north'), attributes={'title': 'Latitude', 'type': 'double', 'valid_max': 90.0, 'valid_min': -90.0}, coord_system=None), dims=(1,), points_dtype=dtype('float64'), bounds_dtype=dtype('float64'), kwargs={'scalar': False, 'circular': False, 'order': 1}), _CoordMetaData(defn=CoordDefn(standard_name='longitude', long_name=None, var_name='lon', units=Unit('degrees_east'), attributes={'title': 'Longitude', 'type': 'double', 'valid_max': 360.0, 'valid_min': 0.0}, coord_system=None), dims=(2,), points_dtype=dtype('float64'), bounds_dtype=dtype('float64'), kwargs={'scalar': False, 'circular': True, 'order': 1})]
```
Note that `bounds_type` is different for latitude and longitude for the two files.

I tried adding the following to `cesm2.py` to overwrite bounds_dtype:
```
            latitude = cube.coord('latitude')
            latitude.bounds_dtype=np.dtype('float64')
            longitude = cube.coord('longitude')
            longitude.bounds_dtype=np.dtype('float64')
```
But that gave the following error:
```
  File "/home/rng/miniconda3/envs/esmvaltool2/lib/python3.7/site-packages/esmvalcore/cmor/_fixes/cmip6/cesm2.py", line 30, in fix_metadata
    latitude.bounds_dtype=np.dtype('float64')
AttributeError: can't set attribute

```
This seems to be because `bounds_dtype` is read-only (see e.g. here https://scitools.org.uk/iris/docs/latest/whatsnew/2.0.html)

[test-cesm.yml.txt](https://github.com/ESMValGroup/ESMValCore/files/4357160/test-cesm.yml.txt)

[main_log_debug.txt](https://github.com/ESMValGroup/ESMValCore/files/4357145/main_log_debug.txt)

- Pointer to existing copy of data :
```
CMIP6/CMIP/NCAR/CESM2/historical/r1i1p1f1/Amon/tas/gn/v20190308/tas_Amon_CESM2_historical_r1i1p1f1_gn_185001-201412.nc
CMIP6/ScenarioMIP/NCAR/CESM2/ssp245/r1i1p1f1/Amon/tas/gn/v20190730/tas_Amon_CESM2_ssp245_r1i1p1f1_gn_201501-206412.nc
```

**Comment by zklaus** [comment]
Thanks for this report! The problem with your fix is that `bounds_dtype` is only a reflection of the actual property of the bounds that you have to change. The following should work instead:
```
            latitude = cube.coord('latitude')
            latitude.bounds = latitude.bounds.astype(np.float64)
            longitude = cube.coord('longitude')
            longitude.bounds = longitude.bounds.astype(np.float64)
```
which actually converts the data type. There is a possibility that the converted bounds are identical to the correct `double` bounds only up to 5 or so decimals, which might prevent concatenation still. Let me know if that happens.

**Comment by npgillett** [comment]
Thanks very much for the response @zklaus! With your fix above in `cesm2.py` the fixes now run OK, but the code fails later in concatenate:
```
  File "/home/rng/miniconda3/envs/esmvaltool2/lib/python3.7/site-packages/esmvalcore/preprocessor/_io.py", line 91, in concatenate
    cube = iris.cube.CubeList(cubes).concatenate_cube()
  File "/home/rng/miniconda3/envs/esmvaltool2/lib/python3.7/site-packages/iris/cube.py", line 511, in concatenate_cube
    raise iris.exceptions.ConcatenateError(msgs)
iris.exceptions.ConcatenateError: failed to concatenate into a single cube.
  An unexpected problem prevented concatenation.
  Expected only a single cube, found 2.
```
Here is the main_log_debug:

[main_log_debug.txt](https://github.com/ESMValGroup/ESMValCore/files/4360980/main_log_debug.txt)

I couldn't figure out how to check whether it was any difference in bounds that was still causing the problem - I found the logic in the iris files hard to follow!

**Comment by zklaus** [comment]
To debug it might be the easiest to in your `config-user.yml` set
```
save_intermediary_cubes: true
remove_preproc_dir: false
```
Then in your output directory in the `preproc` dir you will find one output cube for every step of the preprocessing. Hopefully, then you can see the data after a number of preprocessing steps and trace back from the last one that worked to the point where the problem occurred.

Having said that, I would expect it to be the problem I mentioned above, ie rounding differences due to data type conversion. I would suggest copying the bounds from the correct ones to the faulty ones.

**Comment by npgillett** [comment]
Great - thanks for the response. You were correct @zklaus  - it was a problem with differences in the latitude bounds at > 5 d.p. There may be a more elegant way to do this, but I was able to fix this just by rounding `latitude bounds` to 4 d.p. For the record the CESM2 `tas` files now concatenate successfully with the following fix in `cesm2.py` (I am not an esmvalcore developer, so will leave this to someone else to fold in). 

```
class Tas(Fix):
    """Fixes for tas."""

    def fix_metadata(self, cubes):
        """Fix latitude_bounds.

        Parameters
        ----------
        cube : iris.cube.CubeList

        Returns
        -------
        iris.cube.Cube

         """
        cube = self.get_cube_from_list(cubes)
        add_scalar_height_coord(cube)
        for cube in cubes:
            latitude = cube.coord('latitude')
            latitude.bounds = latitude.bounds.astype(np.float64)
            latitude.bounds=np.round(latitude.bounds,4)
            longitude = cube.coord('longitude')
            longitude.bounds = longitude.bounds.astype(np.float64)

        return cubes

```

**Comment by npgillett** [comment]
@zklaus Sorry for the simple question, but how do I compare the two cubes to see what is different? The other example I mentioned above is NorESM2-LM, which throws the same error when trying to concatenate siconc from historical and ssp245. The error message doesn't tell me what is different about the two cubes, so I don't know what to fix. (I already tried applying the same fix on latitude and longitude bounds but that didn't work there).
[main_log_debug.txt](https://github.com/ESMValGroup/ESMValCore/files/4372235/main_log_debug.txt)
 How can I compare the two cubes to see what is different? Thanks!

**Comment by zklaus** [comment]
@npgillett, I wrote down some tips in the form of FAQ [here](https://github.com/ESMValGroup/ESMValCore/wiki/Debugging-FAQ#comparing-data-with-iris). Let me know if that helps.

**Comment by npgillett** [comment]
@zklaus Thanks - that did help. For some reason I couldn't fully debug this by looking at the NetCDF file, but that page was useful - based on the advice there I used a merge command inserted into the code to identify overlapping time coordinates in some CESM2 hist-GHG and ssp245-GHG files I was trying to concatenate.

**Comment by zklaus** [comment]
Interesting! Ok, I suggest that we finish this issue here and then discuss the other problem in a new issue, ok?

Don't worry about being a developer or not, could you do a pull-request to address this issue with the fix you proposed above?

**Comment by npgillett** [comment]
OK - currently I have esmvaltool installed from conda i.e. not from github. How do I do a pull request?

**Comment by npgillett** [comment]
Here is a fix file to address the above issue, and also to address a second unrelated issue with siconc in CESM2 files such as this (the typesi fix seems to come from an earlier version of esmvaltool but isn't included in the version of esmvalcore currently on the repo):

`DAMIP/NCAR/CESM2/hist-nat/r3i1p1f1/SImon/siconc/gn/v20190730/siconc_SImon_CESM2_hist-nat_r3i1p1f1_gn_185001-189912.nc`
Error message:
```
 File "/home/rng/miniconda3/envs/esmvaltool2/lib/python3.7/site-packages/esmvalcore/cmor/check.py", line 124, in report_errors
    raise CMORCheckError(msg)
esmvalcore.cmor.check.CMORCheckError: There were errors in variable siconc:
typesi: does not exist
in cube:
sea_ice_area_fraction                      (time: 600; cell index along first dimension: 384; cell index along second dimension: 320)

```

[cesm2.py.txt](https://github.com/ESMValGroup/ESMValCore/files/4390493/cesm2.py.txt)

**Comment by rswamina** [comment]
@zklaus and @npgillett - I ran into the same problem when trying to concatenate cubes from the CESM2 model as well. What has been decided on how to fix this?

**Comment by valeriupredoi** [comment]
if it's an issue related to partial time overlaps and multiple experiments to be concatenated try this https://github.com/ESMValGroup/ESMValCore/pull/615

**Comment by mwjury** [comment]
@npgillett @valeriupredoi I ran into the same problem when combining tas Amon historical and ssp585 and here is what I believe to be the main problem which may add to the already identified for NCC  NorESM2-LM:
historical/r1i1p1f1/Amon/tas/gn/v20190815  files:
    lon  [0, 2.5 ... 355, 357.5]
    lon_bnds  [0, 1.25] .... [356.25, 360]
ssp585/r1i1p1f1/Amon/tas/gn/v20191108
    lon  [0, 2.5 ... 355, 357.5]
    lon_bnds  [-1.25, 1.25] .... [356.25, 358.75]
I.e. wrong lon_bnds in the historical.

**Comment by mwjury** [comment]
@npgillett A very similar issue, though not happening in the concatenation of the data is happening for 
/UA/MCM-UA-1-0/historical/r1i1p1f2/Amon/tas/gn/v20190731/tas_Amon_MCM-UA-1-0_historical_r1i1p1f2_gn_185001-201412.nc
UA/MCM-UA-1-0/ssp585/r1i1p1f2/Amon/tas/gn/v20190731/tas_Amon_MCM-UA-1-0_ssp585_r1i1p1f2_gn_201501-210012.nc

again with bad lon_bnds:
longitude [0, 3.75 ... 352.5, 356.25]
lon_bnds [-1.875, 1.875] .... [354.375, 360]

This time they are overlapping and throwing an value error when calling extract_region (cube.intersection):
ValueError: coordinate's range greater than coordinate's unit's modulus

**Comment by mwjury** [comment]
@valeriupredoi @zklaus Can you give me an idea how to approach such issues? Since this is a clear error on the data (description) side, should this be reported to the modellers, or, since it's rather minor issues and is clear where the error lies, be fixed in ESMVal?

**Comment by rswamina** [comment]
@valeriupredoi - Thanks for the fix via the overhaul_concatenation branch. That issue is closed so I am posting this error here. I continue to get an error when concatenating historical and ssp (e.g. ssp126) data for variable "Amon/tas" and ensemble member r1i1p1f1. The error is this:
ERROR   esmvalcore.preprocessor:226 Failed to run concatenate([<iris 'Cube' of air_temperature / (K) (time: 1980; latitude: 192; longitude: 288)>, <iris 'Cube' of air_temperature / (K) (time: 600; latitude: 192; longitude: 288)>, <iris 'Cube' of air_temperature / (K) (time: 432; latitude: 192; longitude: 288)>], {})

followed by (file path information removed):

preprocessor/_io.py" line 172, in _get_concatenation_error
    raise ValueError(f'Can not concatenate cubes: {msg}')
ValueError: Can not concatenate cubes: failed to concatenate into a single cube.
  Dimension coordinates metadata differ: latitude, longitude != latitude, longitude

I don't quite understand why this is different. Please let me know if there is any other information I can provide to help fix this. Thanks!

**Comment by rswamina** [comment]
Sorry I forgot to mention this is with the CESM2 model.

**Comment by rswamina** [comment]
@valeriupredoi  - Can something be done to fix the cube concatenation issue or has this been addressed elsewhere? Thanks!

**Comment by valeriupredoi** [comment]
I am having a look as we speak, let me update yous in a short bit :beer:

**Comment by valeriupredoi** [comment]
OK guys, I looked around - UKESM1 data has no problems concatenating between exps historical and ssp, I can't find any CESM2 ssp data for Amon on Jasmin, (the DRS exists but the actual data doesn't), can you pls give me an updated dataset that has issues with the concatenation? :beer:

**Comment by rswamina** [comment]
@valeriupredoi - I am not sure why this is missing , but the NorESM2 model shares the same grid style and has this problem I believe. NorESM2 data is available on BADC. Meanwhile I'll see if we can get the CESM2 data as well.

**Comment by valeriupredoi** [comment]
right on @rswamina - you totally right, I just ran into the dreaded concat issue from historical+ssp126 with NorESM2-LM data, just in time to call it a day :grin: will investigate thoroughly tomorrow :beer:

**Comment by valeriupredoi** [comment]
OK here's the problem - iris cube concatenation fails because NorESM2-LM has slightly different bounds for `longitude` between the two experiments `historical` and `ssp126`, namely the first and last bounds differ only:
```
hist first and last lon bounds:  [0.   1.25] [356.25 360.  ]
ssp126 first and last lon bounds:  [-1.25  1.25] [356.25 358.75]
```
if we harmonize these two, all goes well fine

**Comment by valeriupredoi** [comment]
created a dedicated issue for the bounds problem https://github.com/ESMValGroup/ESMValCore/issues/806 so am closing this one since the other points here have been fixed, let's take the discussion there :beer:
